ocaml/typing/mode.ml

```diff
@@ -2,1242 +2,1750 @@
-    { locality; uniqueness; linearity }
-
-  let constrain_lower t =
-    let locality = Regionality.constrain_lower t.locality in
-    let uniqueness = Uniqueness.constrain_lower t.uniqueness in
-    let linearity = Linearity.constrain_lower t.linearity in
-    { locality; uniqueness; linearity }
+    let meet (l0, l1, l2) (r0, r1, r2) =
+      ( Locality.Const.meet l0 r0,
+        Linearity.Const.meet l1 r1,
+        Uniqueness.Const.meet l2 r2 )
 
-  let newvar () =
-    let locality = Regionality.newvar () in
-    let uniqueness = Uniqueness.newvar () in
-    let linearity = Linearity.newvar () in
-    { locality; uniqueness; linearity }
-
-  let newvar_below { locality; uniqueness; linearity } =
-    let locality, changed1 = Regionality.newvar_below locality in
-    let uniqueness, changed2 = Uniqueness.newvar_below uniqueness in
-    let linearity, changed3 = Linearity.newvar_below linearity in
-    { locality; uniqueness; linearity }, changed1 || changed2 || changed3
-
-  let newvar_above { locality; uniqueness; linearity } =
-    let locality, changed1 = Regionality.newvar_above locality in
-    let uniqueness, changed2 = Uniqueness.newvar_above uniqueness in
-    let linearity, changed3 = Linearity.newvar_above linearity in
-    { locality; uniqueness; linearity }, changed1 || changed2 || changed3
-
-  let check_const t =
-    let locality = Regionality.check_const t.locality in
-    let uniqueness = Uniqueness.check_const t.uniqueness in
-    let linearity = Linearity.check_const t.linearity in
-    { locality; uniqueness; linearity }
-
-  let print' ?(verbose = true) ppf t =
-    Format.fprintf ppf "%a, %a, %a"
-      (Regionality.print' ~verbose ~label:"locality")
-      t.locality
-      (Uniqueness.print' ~verbose ~label:"uniqueness")
-      t.uniqueness
-      (Linearity.print' ~verbose ~label:"linearity")
-      t.linearity
-
-  let print ppf t = print' ~verbose:true ppf t
+    let join (l0, l1, l2) (r0, r1, r2) =
+      ( Locality.Const.join l0 r0,
+        Linearity.Const.join l1 r1,
+        Uniqueness.Const.join l2 r2 )
+
+    (** constrain uncurried function ret_mode from arg_mode *)
+    let close_over (locality, linearity, uniqueness) =
+      let locality' = locality in
+      (* uniqueness of the returned function is not constrained *)
+      let uniqueness' = Uniqueness.Const.min in
+      let linearity' =
+        Linearity.Const.join linearity
+          (* In addition, unique argument make the returning function once.
+             In other words, if argument <= unique, returning function >= once.
+             That is, returning function >= (dual of argument) *)
+          (Const.unique_to_linear uniqueness)
+      in
+      locality', linearity', uniqueness'
+
+    (** constrain uncurried function ret_mode from the mode of the whole function *)
+    let partial_apply (locality, linearity, _) =
+      let locality' = locality in
+      let uniqueness' = Uniqueness.Const.min in
```

Comment:
The one call site of this function ignores the uniqueness (and uses Shared), as related to a post I made on Slack. This function also chooses the opposite mode from `Oprint.partial_apply`. This should be cleaned up somehow.

Reply:
The call site (in typetexp.ml) interpret user-written types later used to constrain terms. As arrow types cross uniqueness, without proper modal kinds, our current approach is to use `Shared` for right mode.

Reply:
OK. But why does this return `min` when that value is just ignored, always?

Reply:
I guess this is just to keep in sync with the variable version of this function.

Conceptually, since arrow types cross uniqueness, it doesn't matter whether it returns `unique` or `shared`. Currently this is achieved by `typetexp.ml` always using `shared`, in the future this will be achieved by modal kinds.

Reply:
OK. This makes a little more sense now.